packages/opentelemetry-plugin-document-load/src/documentLoad.ts

```diff
@@ -0,0 +1,219 @@
+   * @param entries
+   */
+  private _addSpanEvent(
+    span: Span,
+    performanceName: string,
+    entries: PerformanceEntries
+  ): Span | undefined {
+    if (
+      hasKey(entries, performanceName) &&
+      typeof entries[performanceName] === 'number'
+    ) {
+      span.addEvent(performanceName, undefined, entries[performanceName]);
+      return span;
+    }
+    return undefined;
+  }
+
+  /**
+   * Collects information about performance and creates appropriate spans
+   */
+  private _collectPerformance() {
+    const entries = this._getEntries();
+
+    const rootSpan = this._startSpan(
+      AttributeNames.DOCUMENT_LOAD,
+      PTN.FETCH_START,
+      entries
+    );
+    if (!rootSpan) {
+      return;
+    }
+
+    this._addSpanEvent(rootSpan, PTN.DOMAIN_LOOKUP_START, entries);
+    this._addSpanEvent(rootSpan, PTN.DOMAIN_LOOKUP_END, entries);
+    this._addSpanEvent(rootSpan, PTN.CONNECT_START, entries);
+    this._addSpanEvent(rootSpan, PTN.SECURE_CONNECTION_START, entries);
+    this._addSpanEvent(rootSpan, PTN.CONNECT_END, entries);
+    this._addSpanEvent(rootSpan, PTN.REQUEST_START, entries);
+    this._addSpanEvent(rootSpan, PTN.RESPONSE_START, entries);
+    this._addSpanEvent(rootSpan, PTN.RESPONSE_END, entries);
+    this._addSpanEvent(rootSpan, PTN.UNLOAD_EVENT_START, entries);
+    this._addSpanEvent(rootSpan, PTN.UNLOAD_EVENT_END, entries);
+    this._addSpanEvent(rootSpan, PTN.DOM_INTERACTIVE, entries);
+    this._addSpanEvent(rootSpan, PTN.DOM_CONTENT_LOADED_EVENT_START, entries);
+    this._addSpanEvent(rootSpan, PTN.DOM_CONTENT_LOADED_EVENT_END, entries);
+    this._addSpanEvent(rootSpan, PTN.DOM_COMPLETE, entries);
+
+    this._endSpan(rootSpan, PTN.LOAD_EVENT_START, entries);
+  }
+
+  /**
+   * Helper function for ending span
+   * @param span
+   * @param performanceName name of performance entry for time end
+   * @param entries
+   */
+  private _endSpan(
+    span: Span | undefined,
+    performanceName: string,
+    entries: PerformanceEntries
+  ) {
+    // span can be undefined when entries are missing the certain performance - the span will not be created
+    if (typeof span !== 'undefined' && hasKey(entries, performanceName)) {
+      this._addSpanEvent(span, performanceName, entries);
+      span.end(entries[performanceName]);
+    }
+  }
+
+  /**
+   * gets performance entries of navigation
+   */
+  private _getEntries() {
+    const entries: PerformanceEntries = {};
+    const performanceNavigationTiming = (otperformance.getEntriesByType(
+      'navigation'
+    )[0] as unknown) as PerformanceEntries;
+
+    if (performanceNavigationTiming) {
+      const keys = Object.values(PTN);
+      const startTime = otperformance.timeOrigin;
+      keys.forEach((key: string) => {
+        if (hasKey(performanceNavigationTiming, key)) {
+          const value = performanceNavigationTiming[key];
+          if (typeof value === 'number' && value > 0) {
+            entries[key] = value + startTime;
```

Comment:
Spans use a function `timeInputToHrTime` ([use in Span.ts](https://github.com/open-telemetry/opentelemetry-js/blob/9855636c9ec939775e79559b0fe067224c591e0d/packages/opentelemetry-tracing/src/Span.ts#L61)). That function can take in a `number` and infer that it's a performance.now timestamp ([code link](https://github.com/open-telemetry/opentelemetry-js/blob/9855636c9ec939775e79559b0fe067224c591e0d/packages/opentelemetry-core/src/common/time.ts#L59)). Then it converts it to an `HrTime` tuple in a careful way that makes sure it doesn't lose nanosecond precision ([code to add seconds and nanos separately](https://github.com/open-telemetry/opentelemetry-js/blob/9855636c9ec939775e79559b0fe067224c591e0d/packages/opentelemetry-core/src/common/time.ts#L40)).

If we just add it here, we are losing nanosecond precision because of round-off with the `double` values once you directly add the start time (large epoch-based number) and the event time (small performance now number with higher accuracy).

My suggestion would be to just pass around the performance timings until you actually need it in the span, and then the span will convert it correctly to an HrTime using the utility function. I think the span inputs should accept it fine based on the logic I mentioned above.

What do you think?

Reply:
I'm going add util method to correctly add 2 values that has type of `TimeInput` and then use it here

Reply:
ok ignore my previous comment, I see this is being added automatically now